### DISCUSSION TITLE
[BUG] - Pause at layer pauses before and after the layer?

### DISCUSSION THREAD

**Discussion by AndyHazz** [jschuh/klipper-macros#179] [Q&A] [answered]
Using Cura 5.4.0 and the instructions from here: https://github.com/jschuh/klipper-macros#ultimaker-cura - everything is working pretty well, but I seem to have an issue with the pause at layer script ...

Here's the console output - you can see I wait for the print to start then use the script to tell it to pause at layer 21. It pauses first at 15:26 (I think this was actually layer 20), then I change filament and set it going again (it took me a while, I was in another room), but after one layer (at 16:04) it paused again.

```
16:07  Done printing file
16:06  M117
16:04  Executing scheduled commands at layer 21:
  PAUSE
15:51  SET_HEATER_TEMPERATURE HEATER=extruder TARGET=210
15:51  Preheating extruder to 185
15:51  LOAD_FILAMENT
15:50  G91
  G1 Z+100 F1500
  G90
15:50  G91
  G1 Z+100 F1500
  G90
15:50  G91
  G1 Z+100 F1500
  G90
15:50  SET_HEATER_TEMPERATURE HEATER=extruder TARGET=100
15:49  LOAD_FILAMENT
15:46  Preheating extruder to 205
15:46  UNLOAD_FILAMENT
15:39  Extruder powered down on idle timeout.
15:26  echo: PAUSE, forced by pause_next_layer
15:00  layer 21:
* PAUSE
15:00  pause_at_layer LAYER=21
14:53  Loaded mesh: default
14:53  File selected
```

It seems that first pause, where the console shows 'echo: PAUSE, forced by pause_next_layer' is not supposed to happen? The second pause, where it shows 'Executing scheduled commands at layer 21:  PAUSE' is the only time it should have paused.

I've read the caveats in the Cura setup documentation a few times and although it seems likely they're related to this issue, I can't quite figure out if/why that would have this effect!

**Comment by jschuh** [comment, marked as answer]
The double pause is because you're including both the [mainsail-config](https://github.com/mainsail-crew/mainsail-config) macros and these macros at the same time. (The "PAUSE, forced by pause_next_layer" message is coming from mainsail-config, not these macros).

The [installation instructions](https://github.com/jschuh/klipper-macros#installation) call out this kind of issue:

> **Note:** Make sure you don't currently have any macros that provide the same basic function as the macros in this repository (e.g. the default [Mainsail](https://docs.mainsail.xyz/configuration#macros) or [fluidd](https://docs.fluidd.xyz/configuration/initial_setup#macros) macros). As a rule, you should avoid using multiple sets of macros that override the same base macro (unless you really know what you're doing) because conflicting macros can cause all sorts of weird and frustrating problems.

**Comment by AndyHazz** [comment]
Ahh, of course - I was blind to that one as I'd not added them myself 😳 

So these are all handled (better) or made irrelevant by klipper-macros and need to be removed?

```
[gcode_macro CANCEL_PRINT]
description: Cancel the actual running print

[gcode_macro PAUSE]
description: Pause the actual running print

[gcode_macro RESUME]
description: Resume the actual running print

[gcode_macro SET_PAUSE_NEXT_LAYER]
description: Enable a pause if the next layer is reached

[gcode_macro SET_PAUSE_AT_LAYER]
description: Enable/disable a pause if a given layer number is reached

[gcode_macro SET_PRINT_STATS_INFO]
description: Overwrite, to get pause_next_layer and pause_at_layer feature
  
[gcode_macro _TOOLHEAD_PARK_PAUSE_CANCEL]
description: Helper: park toolhead used in PAUSE and CANCEL_PRINT
  
[gcode_macro _CLIENT_EXTRUDE]
description: Extrudes, if the extruder is hot enough

[gcode_macro _CLIENT_RETRACT]
description: Retracts, if the extruder is hot enough
```
